shared/src/ledger/native_vp/multitoken.rs

```diff
@@ -0,0 +1,656 @@
+//! Native VP for multitokens
+
+use std::collections::{BTreeSet, HashMap};
+
+use thiserror::Error;
+
+use crate::ledger::native_vp::{self, Ctx, NativeVp};
+use crate::ledger::storage;
+use crate::ledger::vp_env::VpEnv;
+use crate::proto::Tx;
+use crate::types::address::{self, Address, InternalAddress};
+use crate::types::storage::{Key, KeySeg};
+use crate::types::token::{
+    is_any_minted_balance_key, is_any_minter_key, is_any_token_balance_key,
+    minter_key, Amount, Change,
+};
+use crate::vm::WasmCacheAccess;
+
+#[allow(missing_docs)]
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("Native VP error: {0}")]
+    NativeVpError(#[from] native_vp::Error),
+}
+
+/// Multitoken functions result
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Multitoken VP
+pub struct MultitokenVp<'a, DB, H, CA>
+where
+    DB: storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: storage::StorageHasher,
+    CA: WasmCacheAccess,
+{
+    /// Context to interact with the host structures.
+    pub ctx: Ctx<'a, DB, H, CA>,
+}
+
+impl<'a, DB, H, CA> NativeVp for MultitokenVp<'a, DB, H, CA>
+where
+    DB: 'static + storage::DB + for<'iter> storage::DBIter<'iter>,
+    H: 'static + storage::StorageHasher,
+    CA: 'static + WasmCacheAccess,
+{
+    type Error = Error;
+
+    const ADDR: InternalAddress = InternalAddress::Multitoken;
+
+    fn validate_tx(
+        &self,
+        _tx: &Tx,
+        keys_changed: &BTreeSet<Key>,
+        verifiers: &BTreeSet<Address>,
+    ) -> Result<bool> {
+        let mut changes = HashMap::new();
+        let mut mints = HashMap::new();
+        for key in keys_changed {
+            if let Some([token, owner]) = is_any_token_balance_key(key) {
+                let pre: Amount = self.ctx.read_pre(key)?.unwrap_or_default();
+                let post: Amount = self.ctx.read_post(key)?.unwrap_or_default();
+                let diff = post.change() - pre.change();
+                if diff.is_negative()
+                    && !(verifiers.contains(owner) || *owner == address::masp())
```

Comment:
You are right. When the owner is the MASP address, the `verifiers` should have the MASP address.
And, for any other owner, the `verifiers` should have the address, too.
This check isn't needed anymore though I moved this from `vp_token`. Thanks.

Reply:
Because that is checked by `vp_user`.
https://github.com/anoma/namada/blob/cd95787afd8dcf4ef4ec3e6b29fa832e30334860/wasm/wasm_source/src/vp_user.rs#L91-L92